crates/compilers/src/preprocessor/deps.rs

```diff
@@ -0,0 +1,352 @@
+        hir: &Hir<'_>,
+        paths: &[PathBuf],
+        src_dir: &PathBuf,
+        root_dir: &Path,
+        mocks: &mut HashSet<PathBuf>,
+    ) -> Self {
+        let mut preprocessed_contracts = BTreeMap::new();
+        let mut referenced_contracts = HashSet::new();
+        for contract_id in hir.contract_ids() {
+            let contract = hir.contract(contract_id);
+            let source = hir.source(contract.source);
+
+            let FileName::Real(path) = &source.file.name else {
+                continue;
+            };
+
+            // Collect dependencies only for tests and scripts.
+            if !paths.contains(path) {
+                let path = path.display();
+                trace!("{path} is not test or script");
+                continue;
+            }
+
+            // Do not collect dependencies for mock contracts. Walk through base contracts and
+            // check if they're from src dir.
+            if contract.linearized_bases.iter().any(|base_contract_id| {
+                let base_contract = hir.contract(*base_contract_id);
+                let FileName::Real(path) = &hir.source(base_contract.source).file.name else {
+                    return false;
+                };
+                path.starts_with(src_dir)
+            }) {
+                // Record mock contracts to be evicted from preprocessed cache.
+                mocks.insert(root_dir.join(path));
+                let path = path.display();
+                trace!("found mock contract {path}");
+                continue;
+            } else {
+                // Make sure current contract is not in list of mocks (could happen when a contract
+                // which used to be a mock is refactored to a non-mock implementation).
+                mocks.remove(&root_dir.join(path));
+            }
+
+            let mut deps_collector = BytecodeDependencyCollector::new(
+                sess.source_map(),
+                hir,
+                source.file.src.as_str(),
+                src_dir,
+            );
+            // Analyze current contract.
+            let _ = deps_collector.walk_contract(contract);
+            // Ignore empty test contracts declared in source files with other contracts.
+            if !deps_collector.dependencies.is_empty() {
+                preprocessed_contracts.insert(contract_id, deps_collector.dependencies);
+            }
+            // Record collected referenced contract ids.
+            referenced_contracts.extend(deps_collector.referenced_contracts);
+        }
+        Self { preprocessed_contracts, referenced_contracts }
+    }
+}
+
+/// Represents a bytecode dependency kind.
+#[derive(Debug)]
+enum BytecodeDependencyKind {
+    /// `type(Contract).creationCode`
+    CreationCode,
+    /// `new Contract`. Holds the name of the contract, args length and call args offset,
+    /// `msg.value` (if any) and salt (if any).
+    New(String, usize, usize, Option<String>, Option<String>),
```

Comment:
easier with `New { name: String, .. }` struct variant

Reply:
updated in https://github.com/foundry-rs/compilers/pull/252/commits/6b907d1abae6446996db900103ec155c7d7f5dd8